internal/notification/handlers/notification_worker.go

```diff
@@ -0,0 +1,494 @@
+	if err == nil {
+		return nil
+	}
+	// if the max attempts are reached, we cancel the notification
+	if event.Sequence() >= uint64(w.config.MaxAttempts) {
+		return w.commands.NotificationCanceled(ctx, tx, event.Aggregate().ID, event.Aggregate().ResourceOwner, err)
+	}
+	// otherwise we retry after a backoff delay
+	return w.commands.NotificationRetryRequested(ctx, tx, event.Aggregate().ID, event.Aggregate().ResourceOwner, notificationEventToRequest(
+		event.Request,
+		event.NotifyUser,
+		w.backOff(event.BackOff),
+	), err)
+}
+
+func (w *NotificationWorker) sendNotification(ctx context.Context, tx *sql.Tx, request notification.Request, notifyUser *query.NotifyUser, e eventstore.Event) error {
+	ctx, err := enrichCtx(ctx, request.TriggeredAtOrigin)
+	if err != nil {
+		err := w.commands.NotificationCanceled(ctx, tx, e.Aggregate().ID, e.Aggregate().ResourceOwner, err)
+		logging.WithFields("instanceID", authz.GetInstance(ctx).InstanceID(), "notification", e.Aggregate().ID).
+			OnError(err).Error("could not cancel notification")
+		return nil
+	}
+
+	// check early that a "sent" handler exists, otherwise we can cancel early
+	sender, ok := sentHandlers[request.EventType]
+	if !ok {
+		err := w.commands.NotificationCanceled(ctx, tx, e.Aggregate().ID, e.Aggregate().ResourceOwner, err)
+		logging.WithFields("instanceID", authz.GetInstance(ctx).InstanceID(), "notification", e.Aggregate().ID).
+			OnError(err).Errorf(`no "sent" handler registered for %s`, request.EventType)
+		return nil
+	}
+
+	var code string
+	if request.Code != nil {
+		code, err = crypto.DecryptString(request.Code, w.queries.UserDataCrypto)
+		if err != nil {
+			return err
+		}
+	}
+
+	colors, err := w.queries.ActiveLabelPolicyByOrg(ctx, request.UserResourceOwner, false)
+	if err != nil {
+		return err
+	}
+
+	translator, err := w.queries.GetTranslatorWithOrgTexts(ctx, request.UserResourceOwner, request.MessageType)
+	if err != nil {
+		return err
+	}
+
+	generatorInfo := new(senders.CodeGeneratorInfo)
+	var notify types.Notify
+	switch request.NotificationType {
+	case domain.NotificationTypeEmail:
+		template, err := w.queries.MailTemplateByOrg(ctx, notifyUser.ResourceOwner, false)
+		if err != nil {
+			return err
+		}
+		notify = types.SendEmail(ctx, w.channels, string(template.Template), translator, notifyUser, colors, e)
+	case domain.NotificationTypeSms:
+		notify = types.SendSMS(ctx, w.channels, translator, notifyUser, colors, e, generatorInfo)
+	}
+
+	args := maps.Clone(request.Args)
+	if len(args) == 0 {
+		args = make(map[string]any)
+	}
+	args["Code"] = code
+	// existing notifications use `OTP` as argument for the code
+	if request.IsOTP {
+		args["OTP"] = code
+	}
+
+	if err := notify(request.URLTemplate, args, request.MessageType, request.UnverifiedNotificationChannel); err != nil {
+		return err
+	}
+	err = w.commands.NotificationSent(ctx, tx, e.Aggregate().ID, e.Aggregate().ResourceOwner)
```

Comment:
I wonder if we can decorate this error returned from `notify` will a "Retryable" or "Fatal" property so we can control whether to throw (and retry) or immediately mark the notification as failed. 

The scenario I'm thinking about is a 4xx error returned from the notification provider that should **never** be retried. An example is the `max_attempts_reached` error that Twilio will return if 5 attempts to send a sms challenge code are requested within 10 minutes (for a single challenge). Another approach might be a property passed to (or returned from) `notify` that can provide back-pressure to the notification worker influencing the retry delay or total attempts. 

I made a hacky solution here: https://github.com/zitadel/zitadel/pull/8936  that shows some of the thinking.

Reply:
I get the intention, but wouldn't a new request to the same phone number result in the same rate limit at Twilio?

Reply:
> but wouldn't a new request to the same phone number result in the same rate limit at Twilio?

Right, we should just never retry that error state. Right? https://www.twilio.com/docs/api/errors/60203

Reply:
Yes. I'll add an error to cancel the notification without a retry similar to your approach and run some tests. 

In the meantime, there's the current image of this PR if you want to already test as well:
ghcr.io/zitadel/zitadel:e51c08ee2af5a9e071973cca7324f7fa03f16300